### PR TITLE
Avoided update CatalogDatabaseCache on the fly.

### PR DESCRIPTION
**Pull request by zuyu**
In distributed version, it is not safe to update cached `CatalogRelationSchema` on the fly, while some storage blocks still have reference to the existing schema.

So, disable updates in the cache for now.

### CODE REVIEW

**Comment by pateljm** [comment]
@zuyu: So what happens when the catalog changes?

File: catalog/CatalogDatabaseCache.cpp [hunk 1]
```diff
@@ -61,15 +61,17 @@ void CatalogDatabaseCache::update(const serialization::CatalogDatabase &proto) {
 
   SpinSharedMutexExclusiveLock<false> lock(relations_mutex_);
   for (int i = 0; i < proto.relations_size(); ++i) {
-    auto relation_schema = make_unique<const CatalogRelationSchema>(proto.relations(i));
-    const relation_id rel_id = relation_schema->getID();
+    const serialization::CatalogRelationSchema &proto_relation = proto.relations(i);
+    const relation_id rel_id = proto_relation.relation_id();
```
Comment (pateljm) [inline catalog/CatalogDatabaseCache.cpp]: Just checking if there is a reason to use relation_id instead of relationID. Not a big deal, but just wanted to know the reason. Thanks!

Comment (hbdeshmukh) [inline catalog/CatalogDatabaseCache.cpp]: I am not sure how much performance intensive this function will be.  However, If performance is a concern we can refactor it as follows:

```
// No protection needed for this -
vector<rel_id> relation_ids;
// populate relation_ids from proto.

vector<relation_id> need_to_add_relation_id

// In a shared lock mode
for each rel_id in relation_ids {
  if (rel_id not found in rel_map_) {
    need_to_add_relation_id.empalce(rel_id)
  }
}
if (!need_to_add_relation_id.empty()) {
// Exclusive lock
// Add the necessary relation IDs to rel_map_.
}
```

**Comment by hbdeshmukh** [comment]
I have one refactoring suggestion for better performance and echoing @pateljm's question on catalog changes.

**Comment by zuyu** [comment]
@pateljm @hbdeshmukh As noted in #110, we does not support schema changes.

When we support schema changes, we may want to have a version number in `CatalogRelationSchema` to indicate whether we need to update `CatalogDatabaseCache`.

If the changed schema has been cached, a naive way is to flush all the blocks belonged to the schema to the disk, before we reset the schema. There are certain scheduling optimization could be done as well, i.e., postpone the schema changes unless no queries are using it.

File: catalog/CatalogDatabaseCache.cpp (hunk 1 above)
Comment (zuyu) [inline catalog/CatalogDatabaseCache.cpp]: @pateljm Do you mean `relation_id` in `proto_relation.relation_id()`?

If so, it is the API generated by protobuf. Thanks!

Comment (zuyu) [inline catalog/CatalogDatabaseCache.cpp]: @hbdeshmukh This function is called once per query, before Foreman sends the first `WorkOrder` for the query.

You're right that we need to mutex protection once we have multiple query support (i.e., once `QueryManager` has merged), and your proposed refactor looks good. So, I prefer to do so in another PR. Thanks!

Comment (hbdeshmukh) [inline catalog/CatalogDatabaseCache.cpp]: I didn't actually think of concurrent queries for the refactoring, but thanks for pointing to that. There are three reasons:

1. You don't need mutex protection to read proto. Therefore line 64 and 65 can be moved out of any protected scope. 
2. Finding whether a rel_id is in rel_map_ only requires shared lock. Therefore line 67 - 71 can be moved to a shared lock scope. 
3. Finally when you want to modify rel_map_, you should grab an exclusive lock. 

I think it is fine to delay the refactoring until we have concurrent query support, but you can add a TODO for that.

Comment (zuyu) [inline catalog/CatalogDatabaseCache.cpp]: @hbdeshmukh Ah, now I got that you were talking about the big `ExclusiveLock`. I wish you could comment on Line 62 regarding this concern.

Actually, I could refactor it in this PR. Thanks!

**Comment by zuyu** [comment]
@hbdeshmukh I've refactored using fine-grained locks to update `CatalogRelationSchema`s. Please take a look. Thanks!

**Comment by hbdeshmukh** [comment]
Looks good, merging.
